internal/xds/translator/authorization.go

```diff
@@ -0,0 +1,266 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package translator
+
+import (
+	"errors"
+	"fmt"
+
+	cncfv3 "github.com/cncf/xds/go/xds/core/v3"
+	matcherv3 "github.com/cncf/xds/go/xds/type/matcher/v3"
+	configv3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	rbacconfigv3 "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	routev3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	rbacv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/rbac/v3"
+	hcmv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	networkinput "github.com/envoyproxy/go-control-plane/envoy/extensions/matching/common_inputs/network/v3"
+	ipmatcherv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/matching/input_matchers/ip/v3"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/envoyproxy/gateway/internal/ir"
+	"github.com/envoyproxy/gateway/internal/xds/types"
+)
+
+func init() {
+	registerHTTPFilter(&rbac{})
+}
+
+type rbac struct{}
+
+var _ httpFilter = &rbac{}
+
+// patchHCM builds and appends the RBAC Filter to the HTTP Connection Manager if
+// applicable.
+func (*rbac) patchHCM(
+	mgr *hcmv3.HttpConnectionManager,
+	irListener *ir.HTTPListener,
+) error {
+	if mgr == nil {
+		return errors.New("hcm is nil")
+	}
+
+	if irListener == nil {
+		return errors.New("ir listener is nil")
+	}
+
+	if !listenerContainsRBAC(irListener) {
+		return nil
+	}
+
+	// Return early if filter already exists.
+	for _, f := range mgr.HttpFilters {
+		if f.Name == string(egv1a1.EnvoyFilterRBAC) {
+			return nil
+		}
+	}
+
+	rbacFilter, err := buildHCMRBACFilter()
+	if err != nil {
+		return err
+	}
+
+	// Ensure the RBAC filter is the first one in the filter chain.
```

Comment:
hmm does this work if we later want to use like JWT user information for the authorization? We might need way to specify the order?

Reply:
It's not what it means. It's a copy-paste typo. 

The original one is "Make sure the router filter is the terminal filter in the chain."

Anyway, we sort all the filters afterward, so the order doesn't mater here.